### PR TITLE
Fixes 5346: Fixed delete from url from preload on 404

### LINKED ISSUE #5346: URLs with error 404, deleted, and trashed posts aren't removed from the wpr_rocket_cache table
**Before submitting an issue please check that you’ve completed the following steps:**
- Made sure you’re on the latest version ✅ `3.12.0.3`
- Used the search feature to ensure that the bug hasn’t been reported before ✅

**Describe the bug**

When:
* the status code of a URL is `404`
* a post of any post type is trashed and deleted

their URLs are not removed from the `wpr_rocket_cache` table. In fact, they are removed and readded with a different ID.

They remain there in an `in-progress` status. When `rocket_preload_revert_old_in_progress_rows` runs the status is changed to `pending` and then `in-progress` again creating a loop.

**To Reproduce**

1. Select a cached post whose status is `completed`.
2. Delete the post and checked the database.
3. It will be there as `pending` with a different `id`.

**Expected behavior**

URLs with an error 404 and posts that are deleted/trashed should be removed from the database.
Undeleting a valid URL should add it back.

**Additional context**

The ` delete_url_on_not_found()` is not currently used:
https://github.com/wp-media/wp-rocket/blob/bb4ab38d0c2f551259bce097d77d7efa4f16fc37/inc/Engine/Preload/Subscriber.php#L201-L205

**Backlog Grooming (for WP Media dev team use only)**
- [x] Reproduce the problem
- [x] Identify the root cause
- [x] Scope a solution
- [x] Estimate the effort

### PR DESCRIPTION
**Pull request by CrochetFeve0251**
## Description

Fix an issue on preload where we don't remove the url from the preload when trashing or deleting it.
For that the `set_404` have been added back.

Fixes #5346

## Type of change

- [ ] Enhancement (non-breaking change which improves an existing functionality)

## Is the solution different from the one proposed during the grooming?

No.

## How Has This Been Tested?

One test has been added to check this functionality is present.

- [ ] Automated Test
- [ ] Manual Testing

# Checklist:

Please delete the options that are not relevant.

- [ ] My code follows the style guidelines of this project
- [ ] I have performed a self-review of my own code
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [ ] My changes generate no new warnings
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] New and existing unit tests pass locally with my changes
- [ ] Any dependent changes have been merged and published in downstream modules

### CODE REVIEW

File: tests/Integration/inc/Engine/Preload/Subscriber/deletePostPreloadCache.php [hunk 1]
```diff
@@ -0,0 +1,63 @@
+<?php
```
Comment (remyperona) [inline tests/Integration/inc/Engine/Preload/Subscriber/deletePostPreloadCache.php]: the test filename should be the same as the method being tested. `deleteUrlOnNotFound.php`

File: tests/Integration/inc/Engine/Preload/Subscriber/deletePostPreloadCache.php [hunk 2]
```diff
@@ -0,0 +1,63 @@
...
+namespace WP_Rocket\Tests\Unit\inc\Engine\Preload\Subscriber;
+
+use WP_Rocket\Tests\Integration\AdminTestCase;
+use WP_Rocket\Tests\Integration\FilterTrait;
+
+/**
+ * @covers \WP_Rocket\Engine\Preload\Subscriber::delete_url_on_not_found
+ * @group  Preload
+ */
+class Test_DeletePostPreloadCache extends AdminTestCase
```
Comment (remyperona) [inline tests/Integration/inc/Engine/Preload/Subscriber/deletePostPreloadCache.php]:
```suggestion
class Test_DeleteUrlOnNotFound extends AdminTestCase
```

**Review by remyperona — CHANGES_REQUESTED** [review]

**Review by remyperona — APPROVED** [review]

**Comment by Mai-Saad** [comment]
@CrochetFeve0251 just noticed that:
1-  if we visited any URL as a logged-in user, the URL will be removed from the table. Can you please check that?

https://user-images.githubusercontent.com/76941962/186695994-1ce9d357-94b6-4f85-91d1-05ad8b5d9a9f.mp4

2- if we already deleted the page permanently before in 3.12.0.3 (it was having _trashed URL), 
after using commit `ec2414281
` and clear and preload, that trashed entry has completed status
Screenshot from 2022-08-25 16-51-27

https://user-images.githubusercontent.com/76941962/186704940-dd6c33f4-225e-486b-8a93-20bfe9a14ed6.mp4

File: tests/Integration/inc/Engine/Preload/Subscriber/deleteUrlOnNotFound.php [hunk 3]
```diff
@@ -0,0 +1,59 @@
...
+	{
+		self::uninstallAll();
+		parent::tear_down_after_class();
+	}
+
+	public function setUp(): void
+	{
+		parent::setUp();
+		add_filter('pre_get_rocket_option_manual_preload', [$this, 'manual_preload']);
+	}
```
Comment (remyperona) [inline tests/Integration/inc/Engine/Preload/Subscriber/deleteUrlOnNotFound.php]:
```suggestion
	public function set_up()
	{
		parent::set_up();
		add_filter('pre_get_rocket_option_manual_preload', [$this, 'manual_preload']);
	}
```

**Review by remyperona — APPROVED** [review]
one minor change in the test

**Review by piotrbak — APPROVED** [review]
Working as expected.

**Pull request merged**